src/api/vault.ts

```diff
@@ -0,0 +1,122 @@
+import { _ludexChallengeApi, ApiConfig } from "./utils";
+
+export enum RedeemType {
+  native = "native", // just native redeem
+  nativeForTokens = "nativeForTokens", // pay tokens to get native
+  tokensForNative = "tokensForNative", // pay native to get tokens
+  tokensForTokens = "tokensForTokens", // pay tokens to get tokens
+}
+
+type BaseRedeemArguments = {
+  user: string;
+  amountRedeemed: string;
+  feeRecipient?: string;
+  gasless?: boolean;
+};
+
+type NativeArguments = {
+  type: "native";
+} & BaseRedeemArguments;
+
+type NativeForTokenArguments = {
+  type: "nativeForTokens";
+  payMint: string;
+  amountGiven: string;
+} & BaseRedeemArguments;
+
+type TokensForNativeArguments = {
+  type: "tokensForNative";
+  receiveMint: string;
+  amountGiven: string;
+} & BaseRedeemArguments;
+
+type TokensForTokensArguments = {
+  type: "tokensForTokens";
+  payMint: string;
+  receiveMint: string;
+  amountGiven: string;
+} & BaseRedeemArguments;
+
+type RedeemArguments =
+  | NativeArguments
+  | NativeForTokenArguments
+  | TokensForNativeArguments
+  | TokensForTokensArguments;
+
+export class VaultAPIClient {
+  ludexChallengeApi: <T>(config: ApiConfig) => Promise<T>;
+  constructor(apiKey: string, baseUrl?: string) {
+    this.ludexChallengeApi = _ludexChallengeApi(apiKey, "vault", baseUrl);
+  }
+
+  async getConstructedTx(args: RedeemArguments) {
+    const params = new URLSearchParams();
+    switch (args.type) {
+      case RedeemType.native: {
+        break;
+      }
+      case RedeemType.nativeForTokens: {
+        if (args.payMint === undefined) {
```

Comment:
does typescript not correctly infer these? it should know that these are not undefined when you check the type

Reply:
it does... but isn't it possible at run time that it fails?